### PR TITLE
Reverse Geocode implementation and many bug fixes and_enanchments for geocoder submodules.

### PR DESCRIPTION
**Pull request by itamair**
Reverse Geocode implementation and many bug fixes and_enanchments for geocoder submodules.

Hi Pol ... here is a new/very IMPORTANT pull request, that fixes many limitations and add meaningful enhancements to the module, mainly in the geocoder_field and geocoder_geofield / geocoder_address submodules.
Sorry I couldn't split all these in single Pull Requests, and even in incremental Commits, because they are somehow interlinked and had some problems with it (don't know why, but Github was basing the Pull Request on an old state of the your actual 8.x-2.x, thus I had to re fork everything form its actual Head).
Anyway I added the synthesis of the main enhanchements embedded in this Pull Request, and can say now the Module is really start working in an amazing way, with the ability to really perform (even) combined Geocode and Reverse Geocode operations on multiple fields of the same entity.
I really invite you to try it out, with the "arcgisonline" plugin (that seems not to need an apy key out of the box), and with the Geofield and the Geofield Map (https://www.drupal.org/project/geofield_map) modules, and even with the Address module (now reallly integrated & working in this PR), both as a source of Geocode and destination of Reverse Geocode from Geofield (this wasn't working at all till now).

Just to better clarifiy and detail the Commits message, bear in mind that this PR add better logics in the Geocoder Fields implementation forcing the following: 
- Geocode method is available for DefaultField (every Dumper) and GeofieldField (forced to wkt Dumper->dump)
- Reverse Geocode method is available for DefaultField (every Dumper) and (new) AddressField (forced to geojson Dumper->dump, postprocessed to array) ONLY if Geofield & Geocoder_Geofield modules are enabled: an existing Geofield is forced to be the source/remote;
- string, text and address field types are unique sources/remote fields for Geocode Operations (not Geofield, because unresonable);
- Geofield fields types are unique sources/remote_fields for Reverse Geocode Operations (because makes sense); 

Besides this what briefly described in the (Mega) Commit message has been implemented.

Everything has been quite widley tested from me (2 - 3 days full time), and I think if merged this might be object of new alpha release.

After this, the next (and final step) I would do is implement the capability to set the Geocoder Plugins options (apikey, ssl, locale, region etc properties) that are completely missing now in the module, and that are really (really) needed to make the Geocoder Plugins somehow working in its backend Geocode/Reverse Geocode operations.
This latter implementation will bring the module state to its first Beta level.

MEGA COMMIT MESSAGE FOR MERGE

- Operative implementation of Reverse Geocode method in the geocoder_field hook_entity_presave;
- Force of better logics for source and destination field types of Geocode and Reverse Geocode capabilities:
- Add a weight element/property (in the DefaultField ContentEntityFormInterface form), and a getOrderedGeocodeFields method, to let the user define Geocode/ReverseGeocode custom order implementation;
- From the DefaultField ContentEntityFormInterface form, added the capability to disable or hide (in the Entity Insert/Edit form) the fields that are object of automatic Geocode or Reverse Geocode;
- Functional integration with the Address module with the add of the Addressfield Plugin, able to be object of Reverse Geocode from a Geofield;
- Added validation constrain and error styling (with css library) on Geocoders plugins population in the GeocoderField and GeocodeFormatter settings forms: cannot be empty;
- Add method to check/fix possible (String/Text) Field and Dumper incompatibilities , that were causing appication crash (field max-length, not UFT-8 text encoding);
- Fix of the Single to Multiple option: is valid only for Geocode (not Reverse Geocode) and will work also with multvalue source field

### CODE REVIEW

File: src/DumperPluginManager.php [hunk 1]
```diff
@@ -20,4 +27,67 @@ public function __construct(\Traversable $namespaces, CacheBackendInterface $cac
...
+   *
+   * @param string $dumper_result
+   *   The Dumper result string.
+   * @param \Drupal\geocoder\DumperInterface|\Drupal\Component\Plugin\PluginInspectionInterface $dumper
+   *   The Dumper.
+   * @param \Drupal\Core\Field\FieldConfigInterface $field_config
+   *   The Field Configuration.
+   */
+  public function fixDumperFieldIncompatibility(&$dumper_result, $dumper, FieldConfigInterface $field_config) {
+
```
Comment (drupol) [inline src/DumperPluginManager.php]: Extraneous empty line.

File: src/DumperPluginManager.php [hunk 2]
```diff
@@ -20,4 +27,67 @@ public function __construct(\Traversable $namespaces, CacheBackendInterface $cac
...
+      $incompatibility_warning_message = t("The '@field_name' field 'max length' property is not compatible with the chosen '@dumper' dumper.<br>Thus <b>be aware</b> <u>the dumper output result has been truncated to @max_length chars (max length)</u>.<br> You are advised to change the '@field_name' field definition or chose another compatible dumper.", [
+        '@field_name' => $field_config->getLabel(),
+        '@dumper' => $dumper->getPluginId(),
+        '@max_length' => $field_config->getFieldStorageDefinition()->getSetting('max_length'),
+      ]);
+
+      $dumper_result = substr($dumper_result, 0, $field_config->getFieldStorageDefinition()->getSetting('max_length'));
+
+      // Display a max-length incompatibility warning message.
+      drupal_set_message($incompatibility_warning_message, 'warning');
```
Comment (drupol) [inline src/DumperPluginManager.php]: I think there is a "messenger service" that we can use now ?

File: modules/geocoder_address/src/Plugin/Geocoder/Field/AddressField.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * @GeocoderField(
+ *   id = "address_field",
+ *   label = @Translation("Address field plugin"),
+ *   field_types = {
+ *     "address"
+ *   }
+ * )
+ */
+class AddressField extends DefaultField {
+
```
Comment (drupol) [inline modules/geocoder_address/src/Plugin/Geocoder/Field/AddressField.php]: Extraneous empty line.

Comment (itamair) [reply modules/geocoder_address/src/Plugin/Geocoder/Field/AddressField.php]: This is perfectly identical to the sibling 
Drupal\geocoder_field\Plugin\Geocoder\Field\DefaultField 
definiton
Green Mark for Php/Drupal Standards in my file from PHPStorm

File: modules/geocoder_address/src/Plugin/Geocoder/Field/AddressField.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      // Add the Option to Reverse Geocode.
+      $element['method']['#options']['destination'] = $this->t('<b>Reverse Geocode</b> from a Geofield type existing field');
+
+      // On Address Field the dumper should always be 'geometry'.
+      $element['dumper'] = [
+        '#type' => 'value',
+        '#value' => 'geojson',
+      ];
+
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_address/src/Plugin/Geocoder/Field/AddressField.php]: Fixed

File: modules/geocoder_field/geocoder_field.module [hunk 5]
```diff
@@ -10,13 +10,48 @@ use Drupal\Core\Entity\EntityInterface;
...
+    }
+  }
+}
 
 /**
  * Implements hook_form_FORM_ID_alter().
  */
 function geocoder_field_form_field_config_edit_form_alter(array &$form, FormStateInterface $form_state, $form_id) {
+  /* @var \Drupal\Core\Entity\EntityForm $entity_form */
+  $entity_form = $form_state->getFormObject();
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/geocoder_field.module]: Fixed

File: modules/geocoder_field/geocoder_field.module [hunk 6]
```diff
@@ -75,73 +113,117 @@ function geocoder_field_entity_presave(EntityInterface $entity) {
...
 
-    $default_values = $entity->get($field_name);
-    $result = [];
-    foreach ($remote_field->getValue() as $delta => $value) {
-      if ($address_collection = \Drupal::service('geocoder')->geocode($value['value'], $geocoder['plugins'])) {
+      foreach ($remote_field->getValue() as $delta => $value) {
         /* @var \Geocoder\Model\AddressCollection $address_collection */
+        switch ($geocoder['method']) {
+          case 'source':
+            $failure_status_message = t("'@text' cannot be Geocoded.", ['@text' => $value['value']]);
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/geocoder_field.module]: Fixed

File: modules/geocoder_field/geocoder_field.module [hunk 7]
```diff
@@ -75,73 +113,117 @@ function geocoder_field_entity_presave(EntityInterface $entity) {
...
-        switch ($geocoder['delta_handling']) {
-
-          // Single-to-multiple handling - if we can, explode out the
-          // component geometries.
-          case 's_to_m':
-            foreach ($address_collection->all() as $collectionDelta => $address) {
-              $result[$collectionDelta] = $dumper->dump($address);
-            }
+          case 'destination':
+            $failure_status_message = t("'@text' cannot be Reverse Geocoded.", ['@text' => $value['value']]);
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/geocoder_field.module]: Fixed

File: modules/geocoder_field/geocoder_field.module [hunk 8]
```diff
@@ -75,73 +113,117 @@ function geocoder_field_entity_presave(EntityInterface $entity) {
...
-          // Default delta handling: just pass one delta to the next.
           default:
-            $result[$delta] = $dumper->dump($address_collection->first());
         }
 
-        continue;
+        if ($address_collection) {
 
+          switch ($geocoder['delta_handling']) {
+
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/geocoder_field.module]: Fixed

File: modules/geocoder_field/geocoder_field.module [hunk 9]
```diff
@@ -75,73 +113,117 @@ function geocoder_field_entity_presave(EntityInterface $entity) {
...
+              // If an Address field is being processed, transform its Dumper
+              // string result into array to comply to Address entity->set Api.
+              if ($geocoder['method'] == 'destination' && $field_config->getType() == 'address' && $dumper->getPluginId() == 'geojson') {
+                $result[$delta] = $dumper_manager->setAddressFieldFromGeojson($result[$delta]);
+              }
+          }
+
+          continue;
+
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/geocoder_field.module]: Fixed

File: modules/geocoder_field/src/GeocoderFieldPluginManager.php [hunk 10]
```diff
@@ -64,14 +64,16 @@ public function __construct(\Traversable $namespaces, CacheBackendInterface $cac
   public function getPluginByFieldType($field_type) {
     foreach ($this->getDefinitions() as $definition) {
       if (in_array($field_type, $definition['field_types'])) {
-        return $this->createInstance($definition['id']);
+        /* @var \Drupal\geocoder_field\GeocoderFieldPluginInterface $geocoder_field_plugin */
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/src/GeocoderFieldPluginManager.php]: The IDE throws a warning if I don't declare @var \Drupal\geocoder_field\GeocoderFieldPluginInterface,
as the createInstance method is returning a generic object.

File: modules/geocoder_field/src/GeocoderFieldPluginManager.php [hunk 11]
```diff
@@ -64,14 +64,16 @@ public function __construct(\Traversable $namespaces, CacheBackendInterface $cac
...
+        $geocoder_field_plugin = $this->createInstance($definition['id']);
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/src/GeocoderFieldPluginManager.php]: Same reason of before

File: modules/geocoder_field/src/GeocoderFieldPluginManager.php [hunk 12]
```diff
@@ -83,16 +85,43 @@ public function getPluginByFieldType($field_type) {
...
+   *   The entity type id.
+   * @param string $bundle
+   *   The bundle.
+   * @param string $field_name
+   *   The field name.
+   *
+   * @return array
+   *   The array of source fields and their label.
+   */
+  public function getReverseGeocodeSourceFields($entity_type_id, $bundle, $field_name) {
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/src/GeocoderFieldPluginManager.php]: created a private method to DRY

File: modules/geocoder_field/src/Plugin/Field/GeocodeFormatterBase.php [hunk 13]
```diff
@@ -246,4 +257,21 @@ public function getEnabledProviderPlugins() {
     return $provider_plugin_ids;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public static function validateGeocodeFormatterPlugins(array $element, FormStateInterface &$form_state) {
+    $plugins = [];
+    $element_values_array = $element['#value'];
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/src/Plugin/Field/GeocodeFormatterBase.php]: Roger

File: modules/geocoder_field/src/PreprocessorPluginManager.php [hunk 14]
```diff
@@ -49,4 +51,56 @@ function ($definition) use ($type) {
...
+    $geocoder_fields = [];
+    $results = [];
+    foreach ($entity->getFields() as $field_name => $field) {
+      /** @var \Drupal\Core\Field\FieldConfigInterface $field_config */
+      if (!($field_config = $field->getFieldDefinition()) instanceof FieldConfigInterface) {
+        // Only configurable fields are subject of geocoding.
+        continue;
+      }
+      $geocoder = $field_config->getThirdPartySettings('geocoder_field');
+      if (empty($geocoder['method']) || $geocoder['method'] === 'none') {
```
[the remark this answers is not in the document]
Comment (itamair) [reply modules/geocoder_field/src/PreprocessorPluginManager.php]: There were already these inconsistencies in the module. The #77 row is identical (has been copied) from the geocoder_field.module (#126)

File: src/DumperPluginManager.php (hunk 1 above)
Comment (itamair) [reply src/DumperPluginManager.php]: Fixed

File: src/DumperPluginManager.php (hunk 2 above)
Comment (itamair) [reply src/DumperPluginManager.php]: Ah right, for testability ... but incredibly it seems not yet: https://www.drupal.org/node/2278383

**Comment by itamair** [comment]
I think I accomplished your requested Changes I agreed with, otherwise wrote my reasons (not to). Please review further/again.

**Pull request closed without merging**